### DISCUSSION TITLE
API Documentation

### DISCUSSION THREAD

**Discussion by amithkoujalgi** [langchain4j/langchain4j#429] [Ideas]
Hello @langchain4j.

First of all, thank you for bringing langchain to the Java world. Its amazing! 🤩

I was wondering if there's a comprehensive API documentation available somewhere for someone to get started with basics (including setup, how and why langchain4j is used, etc) and then progressively navigate through the documentation to explore more and try more advanced stuff.

If its not already available, I was thinking if I could setup a skeleton structure for the documentation using [docusaurus](https://docusaurus.io/). 

I was thinking that we can have well structured comprehensive API documentation such as [this](https://amithkoujalgi.github.io/ollama4j/docs/intro).

Let me know what you think. Thanks!

**Comment by LizeRaes** [comment]
Hi @amithkoujalgi ,
Thank you for your interest and the Ollama4j reference! 
We were planning to continue extending this[ langchain4j documentation website](https://clun.github.io/langchain4j-docs/) in #152 
I wrote an outline for adding code examples and integration lists, I think it would be a great spot to combine this with proper API documentation.

The last days @crutcher has been adding a lot of javadoc as well (thank you so much @crutcher !), let's integrate that there too.

**Reply by amithkoujalgi** [reply]
Thanks @LizeRaes! That's perfect. This is exactly what I was looking for.

Do consider [docusaurus](https://docusaurus.io/) as it is Javascript-oriented and hence can allow a lot of UI customisations and also allows to write up documentation effortlessly.

We could even have a GitHub actions setup such that the API docs content is auto-built and deployed to the site upon code pushes/releases. We could even merge the auto-build to merge Javadoc and the API doc and deploy it to the site together as one artifact. (I have set up the same for Ollama4j 🙂)

Let me know what's your take on this.

And, if there's any help required on documentation, please feel free to reach out to me. Thanks!

**Reply by LizeRaes** [reply]
The docusaurus features sound very interesting, if we want to keep the documentation website up to date, we indeed better have a lot of automated updates coming from the repo 👌
Also, the fact that you already did that awesome documentation for Ollama is very tempting!
Let's however wait some more days to see if @clun reacts in #152 since he has already spent effort on it.
I will reach out soon :)

**Reply by amithkoujalgi** [reply]
Alright @LizeRaes.🙂 Thanks for the update!

**Reply by LizeRaes** [reply]
Hi @amithkoujalgi , since we got no reply from @clun, we decided to move forward with your proposed docusaurus solution. Are you still up for setting up the skeleton? Would be greatly appreciated!
I'm drafting the sitemap in[ this google docs ](https://docs.google.com/document/d/12fF7rlMyswd_PYfviKY2TCmvuy1v-F_XBiXSeVpjnQE/edit?usp=sharing) - if you could already set this up, I will help adding the missing parts afterwards.
Feel free to create a directory langchain4j/doc for the sources :)
Thank you so much!

**Reply by amithkoujalgi** [reply]
Hello @LizeRaes. I'm glad to hear that you considered the solution I proposed! 🤩

Absolutely, I'm excited to get started on setting up the Docusaurus skeleton for `langchain4j`! 

I'll dive into the Google Docs sitemap you've shared and begin the initial setup. If you have any specific preferences or additional details you'd like to include, feel free to let me know.

I'll fork and create a PR with the initial structure, and once the initial structure is in place, you can review it to see if there's something missing. Looking forward to collaborating on this!

Also, I'd like to know what are your thoughts on hosting the docs? Would you be considering using [GitHub pages](https://pages.github.com/) or [Vercel](https://vercel.com/pricing) (free for open-soure projects) or something along those lines? 

I'd personally prefer [GitHub pages](https://pages.github.com/) as is GitHub's own solution that can be enabled for free for public (open-source) projects and usually is meant to host static pages such as documentation/javadoc etc. Looking forward to know your opinion on this.

**Reply by amithkoujalgi** [reply]
I've created a [PR](https://github.com/langchain4j/langchain4j/pull/458) for this @LizeRaes.
Please have a look and do let me know what you think. 🙂 
Thanks!
